### ISSUE TITLE
Unexpected trigger of downstream data tests when the model is used as an argument of those data tests

### ISSUE THREAD

**Issue by Shadowsong27** [fishtown-analytics/dbt#3382]
### Describe the bug

Hello folks.

Say we have 2 models.

view_down
```
models:
  - name: view_down
    columns:
      - name: dw_address_id
         tests:
          - unique
    tests:
      - two_tables_change_count:
          threshold: '50'
          other_model: ref('view_up')
```

and 

view_up
```
models:
  - name: view_up
    columns:
      - name: 'postal_code'
        tests:
          - not_null
```

definition of `two_tables_change_count`, it compares row diff based on conditions

```
{% macro test_two_tables_change_count(model, result_type, dir, threshold, other_model='', dim_raw={}, dim_res={}) %}

with base AS (
   select (
        select count(*)
        from {{ other_model }}
        where 1=1
        {{ generate_cond_clause_from_dict(dim_raw) }}
    ) raw_count,
   (
        select count(*)
        from {{ model }}
        where 1=1
        {{ generate_cond_clause_from_dict(dim_res) }}
   ) result_count
),

{{ generate_two_result_compare_clause(dir, threshold, result_type) }}

{% endmacro %}
```

When executing `data test -m view_up`, it will also trigger the data test of `view_down` - `two_tables_change_count`, which is unexpected to me. 

This is not very desirable for us coz view_up is an upstream task of view_down, and when 
we are following this pattern `run view_up -> test view_up -> run view_down -> test view_down`,
we definitely do not want to have a failed `test view_up` command caused by the failure of one of the `view_down` data tests, since the `run view_down` has not been executed yet.

We have curerntly bypassed this by not declaring an actual model as an
test argument, but rather the model name - which is not very elegant.

so something like this 

```
models:
  - name: view_down
    columns:
      - name: dw_address_id
         tests:
          - unique
    tests:
      - two_tables_change_count:
          threshold: '50'
          other_model: `view_up`
```

and inside the macro:

```
...
with base AS (
   select (
        select count(*)
        from {{ ref(other_model) }}
...
```

My question is:

Is this the intended feature or am I using it wrongly? or is there a command
that could disable the triggering of all downstream healthcheck

### Expected behavior
Not triggering any data tests out side of the model yml definition.

### System information
**Which database are you using dbt with?**
- [ ] postgres
- [X] redshift
- [ ] bigquery
- [ ] snowflake
- [ ] other (specify: ____________)

**The output of `dbt --version`:**
```
installed version: 0.19.1
   latest version: 0.19.1

Up to date!

Plugins:
  - bigquery: 0.19.1
  - snowflake: 0.19.1
  - redshift: 0.19.1
  - postgres: 0.19.1

```

**The operating system you're using:**

MacOS 11.1

**The output of `python --version`:**

3.7.6

**Comment by Shadowsong27** [comment]
Update: the workaround of constructing `ref` in macro was working for a while and then suddenly it stopped working, giving me:

```

  dbt was unable to infer all dependencies for the model "two_tables_change_count_view_address_expand_lte__view_address__counts__0".
  This typically happens when ref() is placed within a conditional block.
  
  To fix this, add the following hint to the top of the model "two_tables_change_count_view_address_expand_lte__view_address__counts__0":
  
  -- depends_on: {{ ref('view_address') }}
  
  > in macro test_two_tables_change_count (macros/test_two_tables_change_count.sql)
  > called by test two_tables_change_count_view_address_expand_lte__view_address__counts__0 (models/view_expand.yml)
  > called by test two_tables_change_count_view_address_expand_lte__view_address__counts__0 (models/view_expand.yml)

```

which makes the `workaround` seemed illegal

**Comment by jtcohen6** [comment]
Hey @Shadowsong27, thanks for the detailed report! I'll try to tease apart some of what's happening here.

The `two_tables_change_count` test directly depends on / selects from _both_ the `view_up` and the `view_down` models. That's true even though it's defined on one of those models and references the other one—they're both considered parents of that test.

The current behavior for [test selection](https://docs.getdbt.com/reference/node-selection/test-selection-examples) is that `dbt test -m model_a` selects _all_ tests that directly depend on / select from `model_a`, even tests that have other parents, too. I agree that this can be confusing! So I actually have some pretty good news.

In v0.20.0, this behavior will change (#2891, #3235) such that `dbt test -m model_a` _only_ selects tests that _only_ depend on `model_a`. Put another way, unless you select both `model_a` and `model_b`, dbt won't run any tests that depends on both `model_a` and `model_b` .

I believe that changes represents the best resolution to this issue. I'll just quickly say that, while I wouldn't recommend the first workaround you shared, it is a pattern that ought to work—there was actually a bug in v0.19.1 associated with this pattern, causing dbt to raise an error about inability to infer some dependencies. We're planning to have a fix for that in v0.19.2. But then, with the change to test selection coming in v0.20.0, I'm hopeful you won't need that workaround at all! :)
